server/oauth.go

```diff
@@ -1,63 +1,77 @@
 package server
 
 import (
-	"context"
 	"fmt"
 	"net/http"
-	"strconv"
+
+	"github.com/google/jsonapi"
+	"github.com/labstack/echo/v4"
 
 	"github.com/bytebase/bytebase/api"
 	"github.com/bytebase/bytebase/common"
 	vcsPlugin "github.com/bytebase/bytebase/plugin/vcs"
-	"github.com/google/jsonapi"
-	"github.com/labstack/echo/v4"
 )
 
 func (s *Server) registerOAuthRoutes(g *echo.Group) {
-	g.POST("/oauth/vcs/:vcsID/exchange-token", func(c echo.Context) error {
-		ctx := context.Background()
-
-		vcsID, err := strconv.Atoi(c.Param("vcsID"))
-		if err != nil {
-			return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Failed to marshal oauth provider's ID: %v", c.Param("id"))).SetInternal(err)
+	g.POST("/oauth/vcs/exchange-token", func(c echo.Context) error {
+		req := &api.ExchangeToken{}
+		if err := jsonapi.UnmarshalPayload(c.Request().Body, req); err != nil {
+			return echo.NewHTTPError(http.StatusBadRequest, "Malformed exchange token request").SetInternal(err)
 		}
-		code := c.Request().Header.Get("code")
 
-		vcs, err := s.store.GetVCSByID(ctx, vcsID)
-		if err != nil {
-			return echo.NewHTTPError(http.StatusInternalServerError, err)
+		var vcsType vcsPlugin.Type
+		var instanceURL string
+		var oauthExchange *common.OAuthExchange
+		if req.ID > 0 {
+			vcs, err := s.store.GetVCSByID(c.Request().Context(), req.ID)
+			if err != nil {
+				return echo.NewHTTPError(http.StatusInternalServerError, err)
+			}
+			if vcs == nil {
+				return echo.NewHTTPError(http.StatusNotFound, fmt.Sprintf("Failed to find VCS, ID: %v", req.ID)).SetInternal(err)
+			}
+
+			vcsType = vcs.Type
+			instanceURL = vcs.InstanceURL
+			oauthExchange = &common.OAuthExchange{
+				ClientID:     vcs.ApplicationID,
+				ClientSecret: vcs.Secret,
+				Code:         req.Code,
+				RedirectURL:  req.Config.RedirectURL,
+			}
+		} else {
+			vcsType = vcsPlugin.Type(c.QueryParam("type"))
+			if vcsType != vcsPlugin.GitLabSelfHost && vcsType != vcsPlugin.GitHubCom {
+				return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Unexpected VCS type: %s", vcsType))
+			}
+
+			instanceURL = req.Config.Endpoint
+			oauthExchange = &common.OAuthExchange{
+				ClientID:     req.Config.ApplicationID,
+				ClientSecret: req.Config.Secret,
+				Code:         req.Code,
+				RedirectURL:  req.Config.RedirectURL,
+			}
 		}
-		if vcs == nil {
-			return echo.NewHTTPError(http.StatusNotFound, fmt.Sprintf("Failed to find VCS, ID: %v", vcsID)).SetInternal(err)
+
+		if oauthExchange.RedirectURL == "" {
```

Comment:
We can get rid of the redirectURL https://sourcegraph.com/github.com/bytebase/bytebase/-/blob/frontend/src/types/oauth.ts?L7

Since the redirectURL is always the same